### ISSUE TITLE
More advanced state control. (url state is not enough) 

### ISSUE THREAD

**Issue by Ingenu1ty** [saltcorn/saltcorn#1932]
Saltcorn needs the ability to control the [whole global state of the app](https://redux.js.org/introduction/core-concepts). I mean centralized storage of global state on the client side, similar to how [redux](https://redux.js.org) does. 

For [example](https://codesandbox.io/s/github/reduxjs/redux/tree/master/examples/shopping-cart):
A user adds items to a cart. The application should be able to instantly display the final price of the cart anywhere on the screen. When changing the set of items in the cart, cart changes should also be instantly recalculated. [Example](https://codesandbox.io/s/github/reduxjs/redux/tree/master/examples/shopping-cart). 

Right now you can only do this if you call the server(using saltcorn's webSockets  capabilities) every time you add (or remove) an item to the cart... But I believe that this kind of task should be done on the client side... or at least the developer should have a choice where to handle such tasks. 

This is just one example (the simplest one) where advanced state control might be needed... So a **client-side global state store would be very useful.**

**Comment by pyhedgehog** [comment]
There are client-side "triggers", where you can write any browser js, but therr are no sane api in this place.

**Comment by Ingenu1ty** [comment]
> There are client-side "triggers", where you can write any browser js

Let's imagine that I use arbitrary js to store the state of a form (for example, a form) somewhere on the browser side (but where exactly??🤯). How can I display this data in show view? Use arbitrary js to directly manipulate the DOM? - sounds like write_whole_app_in_js😵 

My proposal is **quite simple:**

Instead of always binding a view to a table, make it possible to connect a view to a client side store. **Just like tables, but only on the browser side and with reactivity!**

This will open up new horizons for what types of different applications can be created!😎

**Comment by pyhedgehog** [comment]
>  **Just like tables, but only on the browser side and with reactivity!**

This will be good set of plugins:

1. AngularView
2. ReactView
3. etc.

All you need — write them.

**Comment by Ingenu1ty** [comment]
@pyhedgehog 

Is it really possible to do this with plugin creation? My suggestion is to give any existing [view patterns](https://wiki.saltcorn.com/view/ShowPage/view-patterns) the ability to read and write data to this browser_side_pseudoTable store. 

I haven't dug deep into the plugin system, so I'm not sure if it's possible to do this by writing a plugin... Anyway, what I saw looks too complicated for me - I'm not jsSuperHero :)

**Comment by pyhedgehog** [comment]
You've got what I want to say - one must be super-hero to migrate saltcorn from server-side html generation to client-side **retaining compatibility with viewtemplate and fieldview plugins**.
So I'm telling about something like what ColumnsUI is for foobar2000. If you are not using - foobar2000 is windows music player that supports plugins for everything including UI elements, but ColumnUI plugin replaces whole UI and allows *other* UI elements with another API.
One can implement viewtemplate that generates SPA that in its turn calls saltcorn REST API for data manipulations. There will be access issues, but you need not to be superhero to overcome them (it's enough to be "just" a hero).

Now about plugins - I have a dream to implement patch to saltcorn that allows api extension (I mean at least `exports.middleware` with `for(p of getState().plugins)app.use(p.middleware)`) but this is not so easy to implement in tenant-safe way.

PS: Sorry for my English. It seems that things that I want to describe above my knowledge of English constructs so I have to use interlinear translation.

**Comment by christianhugoch** [comment]
Perhaps we can write a small plugin seed. I mean a plugin that is almost empty but has the basic structure.

**Comment by Ingenu1ty** [comment]
@pyhedgehog 
Thanks for the detailed answer👍
But I still think the idea is not so hopeless, and here's why: 

What happens when the url state changes? Saltcorn completely reloads the page? - it looks like not :) So some kind of reactivity of the interface elements is already incorporated into the architecture?... 

What if we (I mean someone who knows how to do this😏) change the code responsible for reacting to url state changes? And this code will respond to both the url state change and the [browser_side_pseudoTable(redux-like)](https://github.com/saltcorn/saltcorn/issues/1932#:~:text=Instead%20of%20always%20binding%20a%20view%20to%20a%20table%2C%20make%20it%20possible%20to%20connect%20a%20view%20to%20a%20client%20side%20store.%20Just%20like%20tables%2C%20but%20only%20on%20the%20browser%20side%20and%20with%20reactivity!) store change. I drew a diagram for a better explanation :) The last block (with green border) is the same in both cases. So there is no need to rewrite components to work in SPA-like mode... 
![imageeee2](https://github.com/saltcorn/saltcorn/assets/138471707/85cbc04d-85c4-4d25-9d03-c60450b461c8)


> PS: Sorry for my English. It seems that things that I want to describe above my knowledge of English constructs so I have to use interlinear translation. 

Its OK. I understand you very well, although I'm not sure if native speakers understand us too.😏 It seems we have the same problem :) we're just like those two guys from [that old joke](https://klaracat.livejournal.com/206059.html#:~:text=%D0%92%D1%81%D1%82%D1%80%D0%B5%D1%87%D0%B0%D1%8E%D1%82%D1%81%D1%8F%20%D0%B4%D0%B2%D0%BE%D0%B5%20%D0%B2%20%D0%9B%D0%BE%D0%BD%D0%B4%D0%BE%D0%BD%D0%B5%3A%0A%2D%20How%20much%20watches%3F%0A%2D%20Six%20clocks.%0A%2D%20Such%20much%3F%0A%2D%20For%20whom%20how...%0A%2D%20MGIMO%20finished%3F%0A%2D%20A%2Da%2Dasking!) 😁 

> Perhaps we can write a small plugin seed. I mean a plugin that is almost empty but has the basic structure. 

If all of the above is not possible, then of course it would be great to have such a plugin... For example, render [react jsx](https://react.dev/learn/writing-markup-with-jsx) with sane api to [redux](https://redux.js.org/) data store...

**Comment by glutamate** [comment]
I agree that something is needed here but not sure what. As a first step i will make it easier to use the Session variable which can be used to access tables by the user's session. The goal is to be able to implement a shopping cart. It should also be possible to transfer a cart from session-ownership to user ownership on login / account creation. I think this will handle some of the use case as you mention but not all

**Comment by glutamate** [comment]
Thanks for pointing out Reka, i will have a look. we are already using Craft.js for the builder

However, we have basic reactivity in Edit views. If you have a field, say age (Integer),  add a container to the Edit view. Under show if... you can enter a formula, e.g. age>35. This will be dynamic & reactive according to the age field. It does not require a round-trip to the server. It should also work if you have a calculated field, it should update dynamically according to form state. Also join fields are dynamic.

We have just that added a similar show if formula to filter, and there was always one in show. For show views this is decided on the server side and same on filter, but it will still appear reactive to the user. Also in Show i think you can now access the state with $ in front of the state variable so you have a bit more flexibility.

Now in edit it is possible that not everything is as reactive compared to the form state as it could be, I think it would be great to go through the various settings and think about where we could introduce field settings that make things more responsive and dynamic according to the form state. But I would like to stay with this model and make it a more complete implementation. For instance I don't think in Edit embedded views that use extra state formula are responding to form state changes.

**Comment by Ingenu1ty** [comment]
@glutamate 
>we are already using Craft.js for the builder 

Yeah, and that's how I found out rekaJS :-) And here's how the author of craftJS [explains why "reka" is the necessary next step.](https://github.com/prevwong/craft.js/issues/507). 

>But I would like to stay with this model and make it a more complete implementation.  

I can see why. You've created an excellent model for creating CRUD applications! And with the latest changes (0.92), it even began to have _some_ SPA-like capabilities, so to speak. But the existing model still does not allow you to create, for example, an application in which: there is a listView on the left, and when you **mouse hover** on the item of the list, several different views change on the right side, etc... RekaJS is just a different model - truly reactive! 

I'm not suggesting abandoning the old model. But adding **rekaView** to it! Like show/edit view but in the following concept: ~~what happens in Vegas - stays in Vegas~~ everything that happened on the browser side stays in the browser **until it is processed explicitly**. I mean something like browser-side events that constantly monitor state* and interact with the server when certain conditions are met.
(*_It’s worth noting here that the word “state” means quite a few different things. Even in Reka, this word refers to both internal state and UI state-props. Here, by state, I always mean the state-props concept, as in React_.) 

The only thing that should distinguish rekaView from just any other plugin is the following mechanism: What is called ["Globals"](https://reka.js.org/docs/concepts/state#:~:text=Additionally%2C%20we%20can%20also%20define%20global%20variables%20that%20can%20be%20referenced) in Reka, must somehow be in the scope in the "extra state formula" in other views. Or "Globals" should have some kind of reflection in the SC state... 

This way we can **get the best of both worlds**: SEO and fast initial page load from server-side rendering, and SPA reactivity from Reka🚀 

Anyway, thanks for your efforts! Merry Christmas, happy holidays, and creative success in the New Year🎄

**Comment by Ingenu1ty** [comment]
Also, features, [like in this issue](https://github.com/saltcorn/saltcorn/issues/2290#issue-2134088624), would be trivial to do with RekaJS model...
